### PR TITLE
[MINI-1731, MINI-4441] In-App Purchase item implementation 

### PR DESCRIPTION
**Pull request by khairul-alam-licon**
# Description
This PR aims to add the following changes in order to support In-App **purchase item** feature in MiniApp SDK.

- (MINI-1731) Added a new module e.g. `in-app-purchase` including an interface e.g. `InAppPurchaseProvider` and several data-classes like JS related to purchasing items.
- (MINI-4441) Implement [Google Play Billing Library](https://developer.android.com/google/play/billing/integrate?authuser=2) in `in-app-purchase` module and provide a default implementation e.g. `InAppPurchaseProviderDefault` to use it in HostApp. It's maybe required to add UI to give random input of product item id.

## Links
- Mini 4441, MINI-1731

# Checklist
- [x] I have read the [contributing guidelines](../.github/CONTRIBUTING.md).
- [x] I have completed the [SDK Development Learning Path](../.github/CONTRIBUTING.md#SDK-Development-Learning-Path) (if submitting a feature PR)
- [x] I wrote/updated tests for new/changed code
- [x] I removed all sensitive data before every commit, including API endpoints and keys
- [x] I ran `./gradlew check` without errors

### CODE REVIEW

**Review by munir-rakuten — APPROVED** [review]
Looks good to me 👍

File: testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/iap/IAPExecutor.kt [hunk 1]
```diff
@@ -0,0 +1,101 @@
+package com.rakuten.tech.mobile.testapp.ui.iap
+
+import android.app.Activity
+import android.util.Log
+import com.android.billingclient.api.*
+import com.rakuten.tech.mobile.testapp.helper.showAlertDialog
+
+class IAPExecutor(val activity: Activity) {
```
Comment (corycaywood) [inline testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/iap/IAPExecutor.kt:8]: Do you think we could set this up similar to the way we setup Ad Mob? I'm worried that the current implementation leaves a lot of duplicate work which needs to be done by the Host App.

For example, we could create a separate module for `in-app-purchase` (similar to the `admob-latest` module) and then create a default implementation class of `InAppPurchaseBridgeDispatcher` which wraps some of the functionality of `IAPExecutor`. Then the Host App just needs to create an instance of the Default InAppPurchaseBridgeDispatcher.

Comment (khairul-alam-licon) [reply testapp/src/main/java/com/rakuten/tech/mobile/testapp/ui/iap/IAPExecutor.kt:8]: Yeah, this is a good idea with having a separate module with those default implementation. I'm looking forward to those changes.

File: in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            }
+        }
+    }
+
+    private fun retrieveSkuDetails() {
+        skuDetails?.let {
+            val billingFlowParams = BillingFlowParams.newBuilder()
+                .setSkuDetails(it)
+                .build()
+            billingClient?.launchBillingFlow(context, billingFlowParams)?.responseCode
```
Comment (corycaywood) [inline in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt:76]: Looks like there needs to be some handling for the `responseCode` here. According to the documentation, normally the response is delivered to the `PurchaseUpdatedListener`, but there are some cases where that listener would not be called and you would instead need to check this responseCode (`ITEM_ALREADY_OWNED`, `ITEM_UNAVAILABLE`, `USER_CANCELED`): https://developer.android.com/reference/com/android/billingclient/api/BillingClient#launchBillingFlow(android.app.Activity,%20com.android.billingclient.api.BillingFlowParams)

So we should at least check here if it's `OK` and then send an `onError` response to the Mini App if not.

**Review by corycaywood — COMMENTED** [review]
@khairul-alam-licon The module setup is looking good, but there are some issues with `InAppPurchaseExcecutor` - see my comments.

One main concern is that it seems like the `skuDetails` won't be available when needed sometimes since they are fetched async, and also I think that `PurchasesUpdatedListener` is where the callbacks should be handled after a purchased is completed. But I'm not sure if I'm completely understanding how the Billing library works, so let me know if you have any questions.

File: miniapp/build.gradle [hunk 3]
```diff
@@ -101,6 +101,9 @@ dependencies {
 
     compileOnly project(':admob-latest')
     testImplementation project(':admob-latest')
+
+    compileOnly project(':in-app-purchase')
+    testImplementation project(':in-app-purchase')
```
[the remark this answers is not in the document]
Comment (munir-rakuten) [reply miniapp/build.gradle]: @corycaywood why do we need those data classes and `InAppPurchaseProvide` in `miniapp`, as we are separating all the inapp purchase related classes in a module? It suppose to be independent of `miniapp` right? Creating a module for only default implementation of `InAppPurchaseProvide`, is that right?

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseProvider.kt [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package com.rakuten.tech.mobile.miniapp.iap
+
+/**
+ * Functionalities related to In-App purchase.
+ */
+interface InAppPurchaseProvider {
```
Comment (munir-rakuten) [inline miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseProvider.kt:6]: `InAppPurchaseProvider` could be renamed By `IAPBridgeDispatcher` as we name all our bridge dispatcher name like this, what do you think?

File: in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+import com.android.billingclient.api.Purchase
+import com.android.billingclient.api.PurchasesUpdatedListener
+import com.android.billingclient.api.SkuDetails
+import com.android.billingclient.api.SkuDetailsParams
+
+/**
+ * The In-App purchase executor which acts as a default implementation of InAppPurchaseProvider.
+ * @param context should use the same activity context for #MiniAppDisplay.getMiniAppView.
+ */
+class InAppPurchaseExecutor(private val context: Activity) : InAppPurchaseProvider {
```
Comment (munir-rakuten) [inline in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt:19]: Rename `InAppPurchaseExecutor` to `IAPBridgeDispatcherDefault` as it is the default implementation of `IAPBridgeDispatcher`

File: in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    override fun purchaseItem(
+        itemId: String,
+        onSuccess: (purchasedProduct: PurchasedProduct) -> Unit,
+        onError: (message: String) -> Unit
+    ) {
+        // purchasing In-App item using GooglePlay billing library
+        startPurchasingProduct(itemId)
+
+        // prepare the product to be invoked using onSuccess
+        skuDetails?.let {
```
[the remark this answers is not in the document]
Comment (khairul-alam-licon) [reply in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt]: Yeah, I have updated the flow for this part.

File: in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+                    this.skuDetails = skuDetails
+                }
+            } else {
+                Log.e(TAG, "There is an error happened while purchasing item")
+            }
+        }
+    }
+
+    private fun retrieveSkuDetails() {
+        skuDetails?.let {
```
[the remark this answers is not in the document]
Comment (khairul-alam-licon) [reply in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt]: Yeah, I have updated this part to ensure the `SkuDetails` are always fetched before launching the billing flow. However, I am still looking for using a coroutine to try the sync method.

File: in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+    private fun handlePurchases(purchase: Purchase) {
+        val params = ConsumeParams.newBuilder().setPurchaseToken(purchase.purchaseToken).build()
+        billingClient?.consumeAsync(params) { billingResult, _ ->
+            when (billingResult.responseCode) {
+                BillingClient.BillingResponseCode.OK -> {
+                    Log.d(TAG, " Update the appropriate dataset")
+                }
+                else -> {
+                    Log.w(TAG, billingResult.debugMessage)
```
[the remark this answers is not in the document]
Comment (khairul-alam-licon) [reply in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt]: Yeah, I have updated this part.

File: in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+import com.android.billingclient.api.Purchase
+import com.android.billingclient.api.PurchasesUpdatedListener
+import com.android.billingclient.api.SkuDetails
+import com.android.billingclient.api.SkuDetailsParams
+
+/**
+ * The In-App purchase executor which acts as a default implementation of InAppPurchaseProvider.
+ * @param context should use the same activity context for #MiniAppDisplay.getMiniAppView.
+ */
+class InAppPurchaseExecutor(private val context: Activity) : InAppPurchaseProvider {
```
[the remark this answers is not in the document]
Comment (khairul-alam-licon) [reply in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt]: Yeah, I have added an interface namely `InAppPurchaseProvider.onEndConnection` for this purpose.

File: in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+    private fun startPurchasingProduct(itemID: String) {
+        setUpBillingClient(itemID)
+        retrieveSkuDetails()
+    }
+
+    private fun setUpBillingClient(productID: String) {
+        billingClient = BillingClient.newBuilder(context)
+            .setListener(purchaseUpdateListener)
+            .enablePendingPurchases()
```
[the remark this answers is not in the document]
Comment (khairul-alam-licon) [reply in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt]: I have tested this part without using `.enablePendingPurchases()` but the `SkuDetails` was not fetching, there is something wrong. Thus, I am considering to keep this addition in Android side.

File: in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt (hunk 2 above)
Comment (khairul-alam-licon) [reply in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt]: Yeah, I have added those responseCode e.g. `ITEM_ALREADY_OWNED`, `ITEM_UNAVAILABLE`, `USER_CANCELED` to be invoked by `onError`.

File: miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseProvider.kt (hunk 4 above)
Comment (khairul-alam-licon) [reply miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseProvider.kt:6]: We already have a `InAppPurchaseBridgeDispatcher` in `miniapp` module to keep some additional setup related to bridge connectivity for this feature. We also have a similar approach like `AdBridgeDispatcher`. 

In this case, I think, it's okay to keep `InAppPurchaseProvider` for this interface.

File: in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt (hunk 5 above)
Comment (khairul-alam-licon) [reply in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt]: Yeah, this name is kinda confusing, so I have renamed to `InAppPurchaseProviderDefault` to be matched with the interface's name.

**Comment by khairul-alam-licon** [comment]
Closing after a discussion with team, we can adopt other scope using a different PR to address this ticket.
Thank you guys for your reviews in this PR.

**Pull request closed without merging**